addons/mail/static/src/service_worker.js

```diff
@@ -39,8 +39,49 @@ self.addEventListener("notificationclick", (event) => {
 });
 self.addEventListener("push", (event) => {
     const notification = event.data.json();
-    self.registration.showNotification(notification.title, notification.options || {});
+    event.waitUntil(handlePushEvent(notification));
 });
+
+/** @type {Number} */
+self.notifDisplayCorrelationId;
+/** @type {Promise} */
+self.notifDisplayResolve;
+/** @type {Number} */
+self.notifDisplayTimeout;
+
+self.addEventListener("message", ({ data }) => {
+    const { type, payload } = data || {};
+    if (
+        type === "notification-display-response" &&
+        payload.correlationId === self.notifDisplayCorrelationId
+    ) {
+        clearTimeout(self.notifDisplayTimeout);
+        self.notifDisplayResolve?.();
+    }
+});
+
+async function handlePushEvent(notification) {
+    const { model, res_id } = notification.options?.data || {};
+    self.notifDisplayCorrelationId = `${Date.now()}-${Math.random().toString(36).slice(2)}`;
+
+    return await new Promise((resolve) => {
+        self.notifDisplayResolve = resolve;
+        self.clients.matchAll({ includeUncontrolled: true, type: "window" }).then((clients) => {
+            clients.forEach((client) => {
+                client.postMessage({
+                    type: "notification-display-request",
+                    payload: { correlationId: self.notifDisplayCorrelationId, model, res_id },
+                });
+                console.log(`Message sent to client: ${client.url}`);
+            });
+        });
+        self.notifDisplayTimeout = setTimeout(() => {
+            resolve(
```

Comment:
It looks like you are going to use the resolved value while its not the case, might be worth it to put it on two lines.